File: planner/app/src/main/java/com/sunilson/firenote/presentation/dialogs/AddElementDialog.java
package com.sunilson.firenote.presentation.dialogs;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.view.WindowManager;
import android.widget.Toast;

import com.google.firebase.database.DatabaseReference;
import com.sunilson.firenote.presentation.application.BaseApplication;
import com.sunilson.firenote.Interfaces.BundleInterface;
import com.sunilson.firenote.Interfaces.MainActivityInterface;
import com.sunilson.firenote.R;
import com.sunilson.firenote.presentation.views.AddElementView;

import javax.inject.Inject;

/**
 * @author Linus Weiss
 */

/**
 * Dialog for adding new Elements
 */
public class AddElementDialog extends SuperDialog {

    private MainActivityInterface mainActivityInterface;
    private AddElementView content;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        super.onCreateDialog(savedInstanceState);

        mainActivityInterface = (MainActivityInterface) ((BaseApplication) getContext().getApplicationContext()).mainContext;

        //Set dialog content
        content = new AddElementView(getActivity(), mainActivityInterface.getSpinnerCategoryAdapter());
        final String elementType = getArguments().getString("elementType");

        //Set dialog title
        titleText.setText(getArguments().getString("title"));
        builder.setCustomTitle(title);

        //Restore state of dialog after orientation change
        if (savedInstanceState != null) {
            content.setColor(savedInstanceState.getInt("color"));
        }

        builder.setView(content);

        //On confirm, create new element
        builder.setPositiveButton(getString(R.string.confirm_add_dialog), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                Element element = null;
                String elementTitle = content.getTitle();

                if (elementType.equals(activity.getString(R.string.element_note))) {
                    element = new Element("note", elementTitle);
                } else if (elementType.equals(activity.getString(R.string.element_checklist))) {
                    element = new Element("checklist", elementTitle);
                } else {
                    element = new Element("bundle", elementTitle);
                }

                element.setColor(content.getColor());

                //Push new element to the correct database reference (depending on if we are in a Bundle or not)
                DatabaseReference dRef = null;
                if (activity instanceof MainActivity) {
                    dRef = mainActivityInterface.getElementsReference().push();
                    element.setCategoryName(content.getCategory().getCategoryName());
                    element.setCategoryID(content.getCategory().getCategoryID());
                    dRef.setValue(element);
                } else {
                    BundleInterface bundleInterface = (BundleInterface) activity;
                    dRef = bundleInterface.getElementsReference().push();
                    element.setCategoryName(content.getCategory().getCategoryName());
                    element.setCategoryID(content.getCategory().getCategoryID());
                    dRef.setValue(element);
                }
                if(!((BaseApplication) getActivity().getApplicationContext()).getInternetConnected()) {
                    Toast.makeText(getActivity(), R.string.edit_no_connection, Toast.LENGTH_LONG).show();
                }
            }
        });

        builder.setNegativeButton(getString(R.string.cancel_add_dialog), new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
            }
        });

        final AlertDialog dialog = builder.create();
        setDialogLayoutParams(dialog);
        dialog.getWindow().getAttributes().windowAnimations = R.style.dialogAnimation;
        return dialog;
    }

    private void setDialogLayoutParams(Dialog dialog) {
        WindowManager.LayoutParams lp = dialog.getWindow().getAttributes();
        lp.width = WindowManager.LayoutParams.MATCH_PARENT;
        lp.height = WindowManager.LayoutParams.MATCH_PARENT;
        dialog.getWindow().setSoftInputMode(
                WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
        dialog.getWindow().setAttributes(lp);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Override
    public void onDestroyView() {
        Dialog dialog = getDialog();

        if (dialog != null && getRetainInstance()) {
            dialog.setDismissMessage(null);
        }
        super.onDestroyView();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("color", content.getColor());
    }

    /**
     * Create new instance of AddElementDialog
     *
     * @param title Title of dialog window
     * @param elementType Element Type that should be created
     * @return New DialogFragment
     */
    public static AddElementDialog newInstance(String title, String elementType) {
        AddElementDialog dialog = new AddElementDialog();
        Bundle args = new Bundle();
        args.putString("title", title);
        args.putString("elementType", elementType);
        dialog.setArguments(args);
        return dialog;
    }
}

File: planner/app/src/main/java/com/sunilson/firenote/presentation/dialogs/AddElementDialog.kt
package com.sunilson.firenote.presentation.dialogs

import android.app.Dialog
import android.os.Bundle
import android.view.WindowManager
import android.widget.Toast
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import com.sunilson.firenote.Interfaces.BundleInterface
import com.sunilson.firenote.R
import com.sunilson.firenote.data.models.Element
import com.sunilson.firenote.presentation.homepage.MainActivity
import com.sunilson.firenote.presentation.shared.base.BaseDialogFragment
import com.sunilson.firenote.presentation.shared.singletons.ConnectivityManager
import com.sunilson.firenote.presentation.views.AddElementView
import javax.inject.Inject

class AddElementDialog : BaseDialogFragment() {

    @Inject
    lateinit var connectivityManager: ConnectivityManager

    private lateinit var addElementView: AddElementView

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        super.onCreateDialog(savedInstanceState)
        addElementView = AddElementView(activity)
        val elementType = arguments?.getString("elementType")
        titleView.dialog_title.text = arguments?.getString("title")
        if(savedInstanceState != null) addElementView.color = savedInstanceState.getInt("color")
        builder.setCustomTitle(titleView)
        builder.setView(addElementView)

        builder.setPositiveButton(getString(R.string.confirm_add_dialog),  { dialog, which ->
            val element = when(elementType) {
                getString(R.string.element_note) -> Element("note", addElementView.title, addElementView.color, addElementView.category)
                getString(R.string.element_checklist) -> Element("checklist", addElementView.title, addElementView.color, addElementView.category)
                else -> Element("bundle", addElementView.title, addElementView.color, addElementView.category)
            }

            if(activity is MainActivity) FirebaseDatabase.getInstance().reference.child("users").child(FirebaseAuth.getInstance().currentUser?.uid).child("elements").child("main").push().setValue(element)
            else (activity as BundleInterface).elementsReference.push().setValue(element)
        })

        builder.setNegativeButton(getString(R.string.cancel_add_dialog), { _, _ ->  dismiss()})

        if(!connectivityManager.isConnected()) Toast.makeText(activity, R.string.edit_no_connection, Toast.LENGTH_LONG).show()

        val dialog = builder.create()
        setDialogLayoutParams(dialog)
        dialog.window.attributes.windowAnimations = R.style.dialogAnimation
        return dialog
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onDestroyView() {
        if (dialog != null && retainInstance) {
            dialog.setDismissMessage(null)
        }
        super.onDestroyView()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt("color", addElementView.color)
    }

    private fun setDialogLayoutParams(dialog: Dialog) {
        val lp = dialog.window!!.attributes
        lp.width = WindowManager.LayoutParams.MATCH_PARENT
        lp.height = WindowManager.LayoutParams.MATCH_PARENT
        dialog.window!!.setSoftInputMode(
                WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)
        dialog.window!!.attributes = lp
    }

    companion object {
        fun newInstance(title: String, elementType: String): AddElementDialog {
            val dialog = AddElementDialog()
            val args = Bundle()
            args.putString("title", title)
            args.putString("elementType", elementType)
            dialog.arguments = args
            return dialog
        }
    }
}